File: protocol/src/main/resources/php/ProtocolManagerTemplate.php
<?php

namespace zfoophp;

include_once 'IProtocolRegistration.php';
${protocol_imports}

class ProtocolManager
{
    private static array $protocols = array();
    private static array $protocolIdMap = array();

    public static function initProtocol(): void
    {
        ${protocol_manager_registrations};
    }

    public static function getProtocolId(mixed $clazz): int
    {
        return self::$protocolIdMap[$clazz];
    }

    public static function getProtocol(int $protocolId): IProtocolRegistration
    {
        return self::$protocols[$protocolId];
    }

    public static function write(ByteBuffer $buffer, mixed $packet): void
    {
        $protocolId = self::getProtocolId($packet::class);
        // write protocol id to buffer
        $buffer->writeShort($protocolId);
        // write packet
        self::getProtocol($protocolId)->write($buffer, $packet);

    }

    public static function read(ByteBuffer $buffer): mixed
    {
        $protocolId = $buffer->readShort();
        return self::getProtocol($protocolId)->read($buffer);
    }
}

File: protocol/src/main/resources/ruby/ProtocolManagerTemplate.rb
${protocol_imports}


class ProtocolManager
	@@protocols = Array.new(32767)
	@@protocolIdMap = Hash.new()

	def self.initProtocol()
		${protocol_manager_registrations}
	end

	def self.getProtocol(protocolId)
		return @@protocols[protocolId]
	end

	def self.write(buffer, packet)
		protocolId = @@protocolIdMap[packet.class]
		buffer.writeShort(protocolId)
		protocol = @@protocols[protocolId]
		protocol.write(buffer, packet)
	end

	def self.read(buffer)
		protocolId = buffer.readShort()
		protocol = @@protocols[protocolId]
		packet = protocol.read(buffer)
		return packet
	end
end